kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h

```diff
@@ -259,6 +259,7 @@ class ResidualBasedNewtonRaphsonStrategy
      
     virtual ~ResidualBasedNewtonRaphsonStrategy()
     {
+        Clear();
```

Comment:
@jcotela I agree that this is an exception to the standard of letting smart pointers do everything automatically. It is however necessary in this case if we want to be able to reuse the ML preconditioner without changing the interface. @RiccardoRossi if by point 1 you mean that the amgcl solver should behave similar to the ML then I don't think this can be fully achieved since the two solvers have different interfaces:

The ML preconditioner is built separately from the solver as

MLPreconditionerPointerType pPrec(new ML_Epetra::MultiLevelPreconditioner(rA, mMLParameterList, true));
Epetra_LinearProblem AztecProblem(&rA,&rX,&rB);
AztecOO aztec_solver(AztecProblem);
aztec_solver.SetPrecOperator(&(*pPrec));

The multigrid structure is built directly in the constructor of the amgcl solver 

auto p_built_in_matrix = amgcl::adapter::zero_copy(A.size1(), A.index1_data().begin(),
                A.index2_data().begin(), A.value_data().begin());

p_solver = boost::make_shared<solver_type>(
                amgcl::adapter::block_matrix<TBlockSize, block_type>(*p_builtin_matrix), prm); 

As I understand, the amgcl interface could be modified to allow storing the solver and reusing it with multiple right hand sides. If this is desired, I think it should be opened as a new issue.